File: native/pe-padding-diversifier/src/pe/address.rs
use serde::{Deserialize, Serialize};

use crate::error::{TransformError, checked_add_usize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
pub struct AddressRange {
  pub start: u32,
  pub len: u32,
}

impl AddressRange {
  pub const fn new(start: u32, len: u32) -> Self {
    Self { start, len }
  }

  pub fn try_from_usize(
    start: usize,
    len: usize,
    context: &'static str,
  ) -> Result<Self, TransformError> {
    checked_add_usize(start, len, context)?;
    let start = u32::try_from(start).map_err(|_| TransformError::IntegerOverflow(context))?;
    let len = u32::try_from(len).map_err(|_| TransformError::IntegerOverflow(context))?;
    Ok(Self { start, len })
  }

  pub fn end(self) -> u32 {
    self.start.saturating_add(self.len)
  }

  pub fn end_u64(self) -> u64 {
    self.start as u64 + self.len as u64
  }

  pub fn contains(self, value: u32) -> bool {
    value as u64 >= self.start as u64 && (value as u64) < self.end_u64()
  }

  pub fn overlaps(self, other: Self) -> bool {
    (self.start as u64) < other.end_u64() && (other.start as u64) < self.end_u64()
  }
}

File: internal/pepad/address.go
package pepad

type addressRange struct {
	Start uint32
	Len   uint32
}

func newRange(start, length uint32) addressRange {
	return addressRange{Start: start, Len: length}
}

func rangeFromUsize(start, length int, context string) (addressRange, error) {
	if _, err := addUsize(start, length, context); err != nil {
		return addressRange{}, err
	}
	if start < 0 || length < 0 || start > int(^uint32(0)) || length > int(^uint32(0)) {
		return addressRange{}, overflow(context)
	}
	return addressRange{Start: uint32(start), Len: uint32(length)}, nil
}

func (r addressRange) end() uint32 {
	return r.Start + r.Len
}

func (r addressRange) endU64() uint64 {
	return uint64(r.Start) + uint64(r.Len)
}

func (r addressRange) contains(value uint32) bool {
	return uint64(value) >= uint64(r.Start) && uint64(value) < r.endU64()
}

func (r addressRange) overlaps(other addressRange) bool {
	return uint64(r.Start) < other.endU64() && uint64(other.Start) < r.endU64()
}

func rangeContainsAny(r addressRange, values []uint32) bool {
	for _, value := range values {
		if r.contains(value) {
			return true
		}
	}
	return false
}

func rangeOverlapsAny(r addressRange, ranges []addressRange) bool {
	for _, other := range ranges {
		if r.overlaps(other) {
			return true
		}
	}
	return false
}
